### PR TITLE
Cloud App Transport

### LINKED ISSUE #1166: SDL 0158 -Cloud App Transport Adapter
Proposal: [Cloud App Transport Adapter](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0158-cloud-app-transport-adapter.md)

> This proposal will detail a possible solution for allowing SDL-enabled cloud applications to be used in a vehicle. This proposal will cover the process of obtaining cloud application server endpoints, enabling which cloud apps should appear in the HMI app list, opening connections between Core and a cloud based app server, and how a cloud application can authenticate connected head units.

Review Issue: smartdevicelink/sdl_evolution#493

Revisions to the review issue: https://github.com/smartdevicelink/sdl_evolution/issues/666

Steering Committee Decision:
> The Steering Committee voted to accept this proposal with revisions. The revisions will include changing the `app_icon_url` policy table attribute to `icon_url` as mentioned in [this comment](https://github.com/smartdevicelink/sdl_evolution/issues/666#issuecomment-459826599), include the affecting platforms based on the protocol changes mentioned in [this comment](https://github.com/smartdevicelink/sdl_evolution/issues/666#issuecomment-460424274), and rename `cloudAppAuthToken` to `authToken` noted in [this comment](https://github.com/smartdevicelink/sdl_evolution/issues/666#issuecomment-460661104).  
Additionally, the author stated that OEMs have the ability to specify secure websockets or unsecured websockets whether you have WS or WSS specified in `cloud_transport_type` in the policy table.

> Revisions have been made

### PR DESCRIPTION
**Pull request by NicoleYarroch**
Fixes #1166 

This PR is **ready** for review.

### Risk
This PR makes **minor** API changes.

### Testing Plan
Test cases added for all new RPCs, structs, and enums

### Summary
* Adds two new RPCS: `SetCloudAppProperties` and `GetCloudAppProperties`
* Adds new parameter, `cloudAppVehicleID`, to the vehicle data RPCS: `SubscribeVehicleData`, `UnsubscribeVehicleData`, `GetVehicleData`, and `OnVehicleData`.
* Adds support for handling an `OnSystemRequest` of type `ICON_URL`. The image data at the provided url is downloaded and the downloaded data is sent to Core via a `SystemRequest` request.
* Adds support for getting the `authToken` from the start service ACK. The functionality has been added to the mobile library, however, Core currently only supports sending an `authToken` to cloud apps. This means that `authToken` will always be `nil` in the library. 
* Bumped RPC spec version to 5.1.0 and protocol spec version to 5.2.0.

##### Breaking Changes
* none

##### Enhancements
* Implemented the Cloud App Transport proposal

##### Bug Fixes
* n/a

### Tasks Remaining:
- [x] Bumped RPC spec version to 5.1.0 and protocol spec version to 5.2.0

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: SmartDeviceLink/SDLManager.h [hunk 1]
```diff
@@ -84,6 +84,11 @@ typedef void (^SDLManagerReadyBlock)(BOOL success, NSError *_Nullable error);
  */
 @property (strong, nonatomic, readonly, nullable) SDLRegisterAppInterfaceResponse *registerResponse;
 
+/**
+ *  The auth token, if received. This should be used to log into a user account.
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLManager.h]: Add "Primarily used for cloud apps with companion app stores."

File: SmartDeviceLink/SDLProxy.m [hunk 2]
```diff
@@ -715,6 +717,27 @@ - (void)handleSystemRequestLockScreenIconURL:(SDLOnSystemRequest *)request {
...
+                    if (error != nil) {
+                        SDLLogW(@"OnSystemRequest (icon url) HTTP download task failed: %@", error.localizedDescription);
+                        return;
+                    } else if (data.length == 0) {
+                        SDLLogW(@"OnSystemRequest (icon url) HTTP download task failed to get the cloud app icon image data");
+                        return;
+                    }
+
+                    SDLSystemRequest *iconURLSystemRequest = [[SDLSystemRequest alloc] initWithType:SDLRequestTypeIconURL fileName:request.url];
+                    iconURLSystemRequest.requestSubType = SDLRPCParameterNameId;
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLProxy.m]: Are you sure this is correct? I don't see it in the proposal.

File: SmartDeviceLink/SDLVersion.h [hunk 3]
```diff
@@ -31,6 +31,7 @@ NS_ASSUME_NONNULL_BEGIN
 - (BOOL)isLessThanVersion:(SDLVersion *)otherVersion;
 - (BOOL)isEqualToVersion:(SDLVersion *)otherVersion;
 - (BOOL)isGreaterThanVersion:(SDLVersion *)otherVersion;
+- (BOOL)isGreaterThanOrEqualVersion:(SDLVersion *)otherVersion;
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLVersion.h]: Change to `isGreaterThanOrEqualToVersion`

May as well also add `isLessThanOrEqualToVersion`

File: SmartDeviceLink/SDLVersion.m [hunk 4]
```diff
@@ -110,6 +110,10 @@ - (BOOL)isGreaterThanVersion:(SDLVersion *)otherVersion {
     return ([self compare:otherVersion] == NSOrderedDescending);
 }
 
+- (BOOL)isGreaterThanOrEqualVersion:(SDLVersion *)otherVersion {
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLVersion.m]: update name based on above

File: SmartDeviceLink/SDLCloudAppProperties.h [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+- (instancetype)initWithAppID:(NSString *)appID NS_DESIGNATED_INITIALIZER;
+
+/**
+ *  Convenience init for all parameters.
+ *
+ *  @param nicknames            An array of app names a cloud app is allowed to register with
+ *  @param appID                The id of the cloud app
+ *  @param enabled              If true, cloud app will be included in HMI RPC UpdateAppList
+ *  @param authToken            Used to authenticate websocket connection on app activation
+ *  @param cloudTransportType   Specifies the connection type Core should use
```
[the remark this answers is not in the document]
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLCloudAppProperties.h]: According to the Core team anything can be entered. Currently the ones that work in Core are `WS` or `WSS`, but an OEM could implement their own transport adapter to handle different values

File: SmartDeviceLink/SDLCloudAppProperties.h [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+ *  @return         A SDLCloudAppProperties object
+ */
+- (instancetype)initWithAppID:(NSString *)appID NS_DESIGNATED_INITIALIZER;
+
+/**
+ *  Convenience init for all parameters.
+ *
+ *  @param nicknames            An array of app names a cloud app is allowed to register with
+ *  @param appID                The id of the cloud app
+ *  @param enabled              If true, cloud app will be included in HMI RPC UpdateAppList
```
[the remark this answers is not in the document]
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLCloudAppProperties.h]: Fixed.

File: SmartDeviceLink/SDLManager.h (hunk 1 above)
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLManager.h]: Fixed. Added

File: SmartDeviceLink/SDLCloudAppProperties.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (joeljfischer) [reply SmartDeviceLink/SDLCloudAppProperties.h]: Mentioning that would be helpful.

File: SmartDeviceLink/SDLHybridAppPreference.h [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+ */
+extern SDLHybridAppPreference const SDLHybridAppPreferenceMobile;
+
+/*
+ *  App preference of cloud.
+ */
+extern SDLHybridAppPreference const SDLHybridAppPreferenceCloud;
+
+/*
+ *  App preference of both mobile and cloud.
```
[the remark this answers is not in the document]
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLHybridAppPreference.h]: Fixed. Clarified the meaning with the Core team

File: SmartDeviceLink/SDLCloudAppProperties.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLCloudAppProperties.h]: Fixed. Added.

File: SmartDeviceLink/SDLProxy.m (hunk 2 above)
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLProxy.m]: Removed. Checked with Core team.

File: SmartDeviceLink/SDLVersion.h (hunk 3 above)
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLVersion.h]: Added

File: SmartDeviceLink/SDLVersion.m (hunk 4 above)
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLVersion.m]: Fixed

**Pull request merged**